### ISSUE TITLE
Security: Should implementations be required to inform users in cases where this is currently recommended?

### ISSUE THREAD

**Issue by ddorwin** [w3c/encrypted-media#312]
The following are from the current Security section:

https://w3c.github.io/encrypted-media/#cdm-security:
>If a user agent chooses to support a Key System implementation that cannot be sufficiently sandboxed or otherwise secured, the user agent SHOULD ensure that users are fully informed and/or give explicit consent before loading or invoking it.

https://w3c.github.io/encrypted-media/#security-prompts:
>User agents SHOULD ensure that users are fully informed and/or give explicit consent before a Key System that presents security concerns that are greater than other user agent features (e.g. DOM content) may be accessed by an origin.

In cases where there are such concerns, should the spec _require_ (`MUST`) fully informing the user? This _might_ address some of the concerns in #304 and related discussion (i.e. in [this thread](https://lists.w3.org/Archives/Public/public-html-media/2016Sep/thread.html#msg3)).

**Comment by mwatson2** [comment]
In both cases, the condition is one that involves judgement. Since "SHOULD" is defined to mean "MUST unless you have a good reason not to" I think it would be close to a noop.

**Comment by ddorwin** [comment]
Yes, this question is whether to require such informing of the user or allow implementers to determine whether they have a good reason not to.

For easy reference, [RFC 2119](https://tools.ietf.org/html/rfc2119) defines these terms as:
>**MUST**   This word, or the terms "REQUIRED" or "SHALL", mean that the
   definition is an absolute requirement of the specification.
...
>**SHOULD**   This word, or the adjective "RECOMMENDED", mean that there
   may exist valid reasons in particular circumstances to ignore a
   particular item, but the full implications must be understood and
   carefully weighed before choosing a different course.

Note that using "SHOULD" means the spec acknowledges "that there may exist valid reasons in particular circumstances to" not inform the user about the greater security concerns.

**Comment by mwatson2** [comment]
SHOULD involves some judgement on the part of the implementor. But my point was that even if you change it to MUST, there is still implementor judgement involved in the conditions that trigger the MUST.

**Comment by ddorwin** [comment]
Ah, yes. I forgot to mention that in the original comment but will cover it in my pending comment.

**Comment by ddorwin** [comment]
#304 requested that EME be disabled by default due to the inherent risks assumed to be presented by any/all CDMs. Stated reasons not to make this change included (paraphrasing):
 1. Not all CDMs pose an increased security risk (or there is no evidence of this)
 1. User fatigue

The spec already defines types of security concerns (i.e. the quotes above), so there is already spec text _attempting_ (see below) to identify a subset of CDMs, which may address (1). Since no user action is (necessarily) required in order to inform the user, (2) should not be an issue.

Thus, _requiring_ that the user be informed _may_ address some of the concerns in #304 and the related Formal Objection.

<br>However, as @mwatson2 mentions, the conditions in quoted from the Security section require implementer judgement. Specifically, there are not clear definitions or tests for:
* "sufficiently sandboxed"
* "otherwise secured"
* "security concerns that are greater than other user agent features (e.g. DOM content)"

Thus, even _requiring_ that the user be informed in these conditions may not have an effect if an implementer judges their solution to be sufficiently secure. (Unless we could find a way to clearly define the conditions.)

However, at least such a change would move the judgement from whether to inform users to whether there is added security risk.

**Comment by steelejoe** [comment]
If the intent of this change is strictly to add another spot at which the implementor considers the security risks for the user, I agree with this change. But if this is going to immediately lead to more discussion of how to define "sufficiently sandboxed", I would say leave this for a future date.

**Comment by ddorwin** [comment]
>If the intent of this change is strictly to add another spot at which the implementor considers the security risks for the user, I agree with this change.

I think it is along these lines - moving the question to "what is the risk to the user?" rather than possibly "do I want to do extra work and add UI)?"

>But if this is going to immediately lead to more discussion of how to define "sufficiently sandboxed", I would say leave this for a future date.

This is certainly related (though perhaps independent of the outcome of this issue), but I don't think we need to solve that in order to address this issue or in v1.

**Comment by ddorwin** [comment]
I created PR #332 for reference.

**Comment by adrianba** [comment]
At some point, every WG gets into a discussion of MUST vs. SHOULD. I think it usually means the work is done. 😃 I think the current language using the word SHOULD is fine in this context since it emphasises @mwatson2's point that there is judgement involved here.

**Comment by jdsmith3000** [comment]
I relate to @adrainba's comment about groups reaching the point where they debate MUST vs. SHOULD.  I could imagine agreeing to this change with discussion, but think it is unnecessary, and in some ways an example of very fine wordsmithing that V1 doesn't really need to be finished.

**Comment by paulbrucecotton** [comment]
@ddorwin - Since we do not have consensus to make this change in EME V1, please change the milestone from V1NonBlocking to VNext.
